File: CrateHacker/collection_search.py
from collections import defaultdict
from thefuzz import fuzz

from app_logging import get_logger
from text_utils import tokenize as _tokenize

log = get_logger(__name__)


def build_collection_index(collection):
    """
    Build an inverted index mapping tokens to collection entry indices.
    Returns (title_index, artist_index) where each maps token -> set of indices.
    """
    title_index = defaultdict(set)
    artist_index = defaultdict(set)

    for i, entry in enumerate(collection):
        for token in _tokenize(entry.get('@TITLE', '')):
            title_index[token].add(i)
        for token in _tokenize(entry.get('@ARTIST', '')):
            artist_index[token].add(i)

    return title_index, artist_index


def _get_candidates(text, index, collection_size):
    """Get candidate entry indices that share at least one token with text."""
    tokens = _tokenize(text)
    if not tokens:
        # No usable tokens — fall back to full scan
        return set(range(collection_size))

    candidates = set()
    for token in tokens:
        candidates.update(index.get(token, set()))
    return candidates


def fuzzy_search(playlist, collection, fuzzy_ratio, title_index=None, artist_index=None,
                 progress_callback=None):
    """
    Search for playlist tracks in the collection using fuzzy matching.

    If title_index/artist_index are provided (from build_collection_index),
    uses token-based pre-filtering to avoid O(n*m) comparisons.

    If progress_callback is provided, it is invoked as
    ``progress_callback(done, total)`` once per processed track. Safe to call
    from a worker thread — keep the callback fast and non-blocking.
    """
    grouped_results = {}
    not_found_tracks = []

    use_index = title_index is not None and artist_index is not None
    collection_size = len(collection)

    total_tracks = len(playlist['items'])
    total_matches = 0
    for i, track in enumerate(playlist['items']):
        spotify_track = track['track']
        artists = ", ".join(item['name'] for item in spotify_track['artists'])

        spotify_key = f"{spotify_track['name']}||{artists}"
        track_title = spotify_track['name'].lower()
        track_artists = artists.lower()

        # Pre-filter candidates using token index, or scan all entries
        if use_index:
            title_candidates = _get_candidates(track_title, title_index, collection_size)
            artist_candidates = _get_candidates(track_artists, artist_index, collection_size)
            # Entries that share tokens with EITHER title or artist
            candidate_indices = title_candidates | artist_candidates
        else:
            candidate_indices = range(collection_size)

        track_matches = []

        for idx in candidate_indices:
            entry = collection[idx]
            entry_title = entry.get('@TITLE', '').lower()
            title_score = fuzz.ratio(track_title, entry_title)
            if (title_score > fuzzy_ratio or
            track_title in entry_title or
            entry_title in track_title):
                try:
                    entry_artists = entry['@ARTIST'].lower()
                except (KeyError, AttributeError):
                    entry_artists = "Unknown"

                artist_score = fuzz.ratio(track_artists, entry_artists)
                if (artist_score > fuzzy_ratio or
                track_artists in entry_artists or
                entry_artists in track_artists):
                    combined_score = (title_score + artist_score) // 2
                    track_matches.append({'entry': entry, 'score': combined_score})

        if track_matches:
            track_matches.sort(key=lambda m: m['score'], reverse=True)
            grouped_results[spotify_key] = {
                'spotify_track': spotify_track,
                'spotify_artists': artists,
                'collection_matches': track_matches
            }
            total_matches += len(track_matches)
        else:
            # Use a clean "Artist - Title" form so downstream disk-search
            # fuzzy matching and Traktor ENTRY generation can parse it
            # without stripping a "Track not found:" prefix.
            not_found_tracks.append(f"{artists} - {spotify_track['name']}")

        if progress_callback is not None:
            try:
                progress_callback(i + 1, total_tracks)
            except Exception as cb_err:
                # Never let a UI callback take down the search
                log.warning("progress_callback raised (ignored): %s", cb_err)

    log.info("Fuzzy search complete: %d matches across %d/%d playlist tracks",
             total_matches, len(grouped_results), total_tracks)

    return grouped_results, not_found_tracks

File: web/src/collectionSearch.ts
/**
 * Fuzzy match Spotify playlist tracks against a Traktor collection.
 * Mirrors CrateHacker/collection_search.py — token-indexed pre-filter,
 * fuzz.ratio() (via fuzzball) + substring containment on both title and
 * artist, combined score = floor((title + artist) / 2).
 */
import { ratio as fuzzRatio } from 'fuzzball';

import type { NmlEntry } from './nml';
import { tokenize } from './tokenize';

export type TokenIndex = Map<string, Set<number>>;

export interface SpotifyArtist {
  name: string;
}

export interface SpotifyTrack {
  name: string;
  artists: SpotifyArtist[];
}

export interface PlaylistItem {
  track: SpotifyTrack;
}

export interface Playlist {
  items: PlaylistItem[];
}

export interface CollectionMatch {
  entry: NmlEntry;
  score: number;
}

export interface GroupedResult {
  spotify_track: SpotifyTrack;
  spotify_artists: string;
  collection_matches: CollectionMatch[];
}

export type ProgressCallback = (done: number, total: number) => void;

export interface FuzzySearchOptions {
  titleIndex?: TokenIndex;
  artistIndex?: TokenIndex;
  progressCallback?: ProgressCallback;
}

export interface FuzzySearchResult {
  groupedResults: Map<string, GroupedResult>;
  notFoundTracks: string[];
}

function addToIndex(map: TokenIndex, key: string, value: number): void {
  let bucket = map.get(key);
  if (!bucket) {
    bucket = new Set();
    map.set(key, bucket);
  }
  bucket.add(value);
}

function readString(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

/**
 * Build inverted indexes mapping tokens → sets of collection-entry indices,
 * one for `@TITLE` and one for `@ARTIST`.
 */
export function buildCollectionIndex(collection: NmlEntry[]): {
  titleIndex: TokenIndex;
  artistIndex: TokenIndex;
} {
  const titleIndex: TokenIndex = new Map();
  const artistIndex: TokenIndex = new Map();

  collection.forEach((entry, i) => {
    for (const tok of tokenize(readString(entry['@TITLE']))) {
      addToIndex(titleIndex, tok, i);
    }
    for (const tok of tokenize(readString(entry['@ARTIST']))) {
      addToIndex(artistIndex, tok, i);
    }
  });

  return { titleIndex, artistIndex };
}

function getCandidates(text: string, index: TokenIndex, collectionSize: number): Set<number> {
  const tokens = tokenize(text);
  const result = new Set<number>();
  if (tokens.size === 0) {
    // No usable tokens — fall back to full scan.
    for (let i = 0; i < collectionSize; i += 1) result.add(i);
    return result;
  }
  for (const tok of tokens) {
    const hits = index.get(tok);
    if (hits) {
      for (const idx of hits) result.add(idx);
    }
  }
  return result;
}

/**
 * Run fuzzy matching for every track in `playlist` against `collection`.
 *
 * When `titleIndex` + `artistIndex` are supplied, only entries that share at
 * least one token with the track's title OR artist are scored — this is the
 * O(n·k) fast path. Without the indexes the function falls back to an O(n·m)
 * full scan, useful for tiny collections and tests.
 *
 * `progressCallback(done, total)` is invoked once per processed track. Errors
 * thrown from the callback are swallowed so a flaky UI listener can't abort
 * the search.
 */
export function fuzzySearch(
  playlist: Playlist,
  collection: NmlEntry[],
  fuzzyRatio: number,
  options: FuzzySearchOptions = {},
): FuzzySearchResult {
  const { titleIndex, artistIndex, progressCallback } = options;
  const useIndex = titleIndex !== undefined && artistIndex !== undefined;
  const collectionSize = collection.length;

  const groupedResults = new Map<string, GroupedResult>();
  const notFoundTracks: string[] = [];

  const totalTracks = playlist.items.length;

  playlist.items.forEach((item, i) => {
    const spotifyTrack = item.track;
    const artists = spotifyTrack.artists.map((a) => a.name).join(', ');
    const spotifyKey = `${spotifyTrack.name}||${artists}`;
    const trackTitle = spotifyTrack.name.toLowerCase();
    const trackArtists = artists.toLowerCase();

    let candidateIndices: Iterable<number>;
    if (useIndex) {
      const titleCands = getCandidates(trackTitle, titleIndex, collectionSize);
      const artistCands = getCandidates(trackArtists, artistIndex, collectionSize);
      const merged = new Set<number>(titleCands);
      for (const idx of artistCands) merged.add(idx);
      candidateIndices = merged;
    } else {
      const all: number[] = [];
      for (let k = 0; k < collectionSize; k += 1) all.push(k);
      candidateIndices = all;
    }

    const trackMatches: CollectionMatch[] = [];
    for (const idx of candidateIndices) {
      const entry = collection[idx];
      if (!entry) continue;
      const entryTitle = readString(entry['@TITLE']).toLowerCase();
      const titleScore = fuzzRatio(trackTitle, entryTitle);
      if (
        titleScore > fuzzyRatio ||
        entryTitle.includes(trackTitle) ||
        trackTitle.includes(entryTitle)
      ) {
        // Mirror the Python branch: when @ARTIST is missing the comparison
        // string is the literal "Unknown" (capital U, not lowercased).
        const rawArtist = entry['@ARTIST'];
        const entryArtists = typeof rawArtist === 'string' ? rawArtist.toLowerCase() : 'Unknown';
        const artistScore = fuzzRatio(trackArtists, entryArtists);
        if (
          artistScore > fuzzyRatio ||
          entryArtists.includes(trackArtists) ||
          trackArtists.includes(entryArtists)
        ) {
          trackMatches.push({ entry, score: Math.floor((titleScore + artistScore) / 2) });
        }
      }
    }

    if (trackMatches.length > 0) {
      trackMatches.sort((a, b) => b.score - a.score);
      groupedResults.set(spotifyKey, {
        spotify_track: spotifyTrack,
        spotify_artists: artists,
        collection_matches: trackMatches,
      });
    } else {
      // Clean "Artist - Title" so downstream disk search can parse it without
      // having to strip a "Track not found: " prefix.
      notFoundTracks.push(`${artists} - ${spotifyTrack.name}`);
    }

    if (progressCallback) {
      try {
        progressCallback(i + 1, totalTracks);
      } catch {
        // Never let a UI listener take down the search.
      }
    }
  });

  return { groupedResults, notFoundTracks };
}
